Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,376 @@
+  declareProperty("CreateIntermediateWorkspaces", false, "Output workspaces showing intermediate working steps");
+  declareProperty("WorkspaceIndex", EMPTY_INT(), m_validators["mustBeNonNegative"],
+                  "If set, only this spectrum will be searched for peaks "
+                  "(otherwise all are)");
+  declareProperty("EstimatedPeakExtent", EMPTY_DBL(), m_validators["mustBeGreaterThanZero"],
+                  "Estimated PeakExtent of the peaks to be found");
+  declareProperty("EstimatedPeakExtentNBins", EMPTY_INT(), m_validators["mustBeGreaterThanOne"],
+                  "Optional: Estimated PeakExtent of the peaks to be found in number of bins");
+  declareProperty("IOversigmaThreshold", 2.0, m_validators["mustBeGreaterThanZero"],
+                  "Minimum Signal/Noise ratio for a peak to be considered significant");
+  declareProperty("PerformBinaryClosing", true,
+                  "Attempt to remove inflections in the data, where a local"
+                  " minima/maxima occurs which is not signficiant enough to be considered a peak");
+  declareProperty("FindHighestDataPointInPeak", false,
+                  "When searching for peaks in the raw data around the iOverSigma maxima, take the highest value,"
+                  " rather than favouring datapoints closer to the maxima");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void FindPeaksConvolve::exec() {
+  storeClassProperties();
+  auto validation_res = secondaryValidation();
+  if (validation_res.second != 0) {
+    throw std::invalid_argument("Validation error: " + validation_res.first);
+  }
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputDataWS))
+  for (int i = 0; i < static_cast<int>(m_specCount); i++) {
+    performConvolution(i);
+  }
+  outputResults();
+}
+
+void FindPeaksConvolve::initiateValidators() {
+  m_validators.reserve(2); // Update depending upon number of validators.
+
+  auto mustBeNonNegative{std::make_shared<Kernel::BoundedValidator<int>>()};
+  mustBeNonNegative->setLower(0);
+  m_validators.emplace(std::make_pair("mustBeNonNegative", std::move(mustBeNonNegative)));
+
+  auto mustBeGreaterThanZero{std::make_shared<Kernel::BoundedValidator<double>>()};
+  mustBeGreaterThanZero->setLowerExclusive(0.0);
+  m_validators.emplace("mustBeGreaterThanZero", std::move(mustBeGreaterThanZero));
+
+  auto mustBeGreaterThanOne{std::make_shared<Kernel::BoundedValidator<int>>()};
+  mustBeGreaterThanOne->setLowerExclusive(1);
+  m_validators.emplace("mustBeGreaterThanOne", std::move(mustBeGreaterThanOne));
+}
+
+std::pair<std::string, int> FindPeaksConvolve::secondaryValidation() const {
+  std::string err_str{};
+  int err_code{0};
+
+  err_str += validatePeakExtentInput();
+
+  if (err_str != "") {
+    err_code = 1;
+  }
+  return std::make_pair(err_str, err_code);
+}
+
+const std::string FindPeaksConvolve::validatePeakExtentInput() const {
```

Comment:
I've tested this, and due to the padding it doesn't matter what the peak extent is. The more bins in the kernel, the longer the convolutions will take, so this is why it appears to hang.

Because of this, I've added a warning if peak extent is greater than the x range, rather than erroring.

Reply:
It may work but it doesn't necessarily make sense! The convolution isn't valid also physically if the kernel size is larger than the extent of the data then the background shell goes outside the data and it won't actually be returning the I/sigma. Therefore I think it should be stopped in the validation!